NewApp/ngsa-csharp/Ngsa.Middleware/NgsaLog.cs

```diff
@@ -0,0 +1,174 @@
+                Console.WriteLine(JsonSerializer.Serialize(d, Options));
+                Console.ResetColor();
+            }
+        }
+
+        public void LogWarning(string message)
+        {
+            const LogLevel logLevel = LogLevel.Warning;
+
+            if (LogLevel > logLevel)
+            {
+                return;
+            }
+
+            Dictionary<string, object> d = GetDictionary(message, logLevel);
+
+            Console.ForegroundColor = ConsoleColor.Yellow;
+            Console.WriteLine(JsonSerializer.Serialize(d, Options));
+            Console.ResetColor();
+        }
+
+        public void LogError(string message, Exception ex = null)
+        {
+            const LogLevel logLevel = LogLevel.Error;
+
+            if (LogLevel > logLevel)
+            {
+                return;
+            }
+
+            Dictionary<string, object> d = GetDictionary(message, logLevel);
+
+            if (ex != null)
+            {
+                d.Add("ExceptionType", ex.GetType());
+                d.Add("ExceptionMessage", ex.Message);
+
+                //d.Add("exceptionText", ex.ToString());
+            }
+            else if (Exception != null)
+            {
+                d.Add("ExceptionType", Exception.GetType().FullName);
+                d.Add("ExceptionMessage", Exception.Message);
+
+                //d.Add("exceptionText", Exception.ToString());
+            }
+
+            // display the error
+            Console.ForegroundColor = ConsoleColor.Red;
+            Console.Error.WriteLine(JsonSerializer.Serialize(d, Options));
+            Console.ResetColor();
+        }
+
+        private Dictionary<string, object> GetDictionary(string message, LogLevel logLevel)
+        {
+            Dictionary<string, object> data = new Dictionary<string, object>
+            {
+                { "Date", DateTime.UtcNow },
+                { "LogName", Name },
+                { "Method", Method },
+                { "Message", message },
+                { "LogLevel", logLevel.ToString() },
+            };
+
+            if (EventId.Id > 0)
+            {
+                data.Add("EventId", EventId.Id);
+            }
+
+            if (!string.IsNullOrWhiteSpace(EventId.Name))
+            {
+                data.Add("EventName", EventId.Name);
+            }
+
+            if (Context != null && Context.Items != null)
+            {
+                data.Add("Path", Context.Request.Path + (string.IsNullOrWhiteSpace(Context.Request.QueryString.Value) ? string.Empty : Context.Request.QueryString.Value));
+
+                if (Context.Items != null)
+                {
+                    CorrelationVector cv = Middleware.CorrelationVectorExtensions.GetCorrelationVectorFromContext(Context);
+
+                    if (cv != null)
+                    {
+                        data.Add("CVector", cv.Value);
+                    }
+                }
+            }
+
+            if (Data != null)
```

Comment:
Data is initialized and won't be null: ` public Dictionary<string, string> Data { get; } = new Dictionary<string, string>();`

So, I think the null check is unnecessary unless Data is possibly set to null somewhere else in the code.  Could make it read only which would mean that it can't be assigned a null value, but using the methods of the dictionary you can still modify the contents.

Reply:
fixed - Data is read-only outside the constructor